[Team 18-Android] 화면 네비게이션 + 달력 구현 + 요금 슬라이더 구현

taewooyo opened the pull request:
안녕하세요 앨런😀   2주차 첫번째 피드백을 받게 된 @bn-tw2020 @shncder 입니다.
제공해주신 깃허브 샘플 레포지토리를 통해 프로젝트 네비게이션을 수정했습니다.

개발하는 입장에서도 보다 편해진 것 같고 좋은 것 같습니다. compose을 처음 학습하면서 사용하다보니 따끔한 피드백 부탁드리겠습니다.

===========================================================================================   

### 📑 구현 

#### 📮 피드백

- 전체적인 화면에 대한 틀 구성 후, navigation 설정

- 달력 구현
  - 커스텀 캘린더 구현
  - List으로 날짜를 관리하여 LazyRow, LazyColumn을 통해 그려주기
  - 클릭을 했는지 안했는지에 대해 enum class로 상태 관리

- 요금 슬라이더 구현
  - Padding으로 간격을 두어 금액에 맞는 곳의 숙소 개수 * 30으로 그려주기
  - RangeSlider을 통해 가격으로 실시간 반응 처리

|검색 리스트 화면(검색전)|검색 리스트 화면(검색후)|날짜 선택 화면|
|-|-|-|
|<img src="https://user-images.githubusercontent.com/91953080/171132163-806cbea1-23e2-48f3-9617-3f6fa373bf2b.png" width="300" height="600">|<img src="https://user-images.githubusercontent.com/91953080/171132175-56941ff4-0057-4723-9cff-fdfc39e20915.png" width="300" height="600">|<img src="https://user-images.githubusercontent.com/91953080/171132194-5905864f-4407-4908-96db-281f2d5799c1.png" width="300" height="600">|

### 🤔 질문사항

- for문으로 그래프를 그린뒤 그래프가 거꾸로 나타나는 현상이 있었는데 현재는 Box를 graphicsLayer를 사용해 뒤집는 방법으로 해결은 해놓았는데 왜 뒤집혀서 나왔는지 알수가 없었습니다.

<img src="https://user-images.githubusercontent.com/91953080/171133907-81223e80-2f1d-457f-9766-92975aba8205.png" width="300" height="600">

taewooyo committed:
feat: 레이아웃 전체 구성(25)

레이아웃 전체 구성
- 각 화면별 레이아웃 구조 설정
- navigation을 이용해 화면 이동

taewooyo committed:
chore: 패키지 위치 변경(25)

taewooyo committed:
feat: 예약 날짜 달력 구현(20)

현재 해 달력 커스텀 마이징
- LocaleDate을 통해 월별 list로 관리

shncder committed:
Merge pull request #26 from bn-tw2020/featrue/layout-25

[feat] 레이아웃 전체 구성 및 달력 구현(20, 25)

shncder committed:
feat: 요금화면 추가(28)

taewooyo committed:
Merge pull request #29 from bn-tw2020/feature/fareScreen

[feat] 요금화면 추가(28)

taewooyo committed:
chore: 함수 분리 및 데이터 전달

검색을 위한 데이터 전달
- 지역, 날짜, 인원 수를 저장을 위해 searchViewModel 공유로 관리

함수 분리
- FareScreen 복잡한 로직을 분리 처리

sabgilhun commented on app/src/main/java/com/example/todo/airbnb/data/datasource/DateLocalDataSource.kt:
```diff
@@ -0,0 +1,61 @@
+
+typealias CalendarYear = List<CalendarMonth>
+
+@RequiresApi(Build.VERSION_CODES.O)
```
`@RequiresApi` 어노테이션은 해당 어노테이션이 붙은 class, method, field 등이 명시한 버전에서만 호출 된다고 표시하는 것 입니다.

하지만 현재 프로젝트는 `minSdk 21`로 정의 해두었기 때문에 ` 26 보다 낮음으로 하위 버전 기종에선 `LocalDate.of` 호출할때 크래쉬가 날거에요.

`@RequiresApi`은 컴파일 에러를 무시해줄뿐 런타임 에러는 날 수 있다는 걸 조심해주세요!

해결 방법은 아예 `LocalDate` 안쓰거나, minSdk 올리던가 아니면 아래 디슈가링 방법으로 해결 가능합니다.
https://developer.android.com/studio/write/java8-support?hl=ko#library-desugaring

taewooyo replied:
LocalDate를 안쓰면 Date나 Calendar을 사용하면 해결할 수 있을 것 같습니다.
minSdk는 최소버전에 대해서 gradle에서 버전을 설정할 수 있을 것 같습니다.
디슈가링에 대해서는 처음 들어보고 학습하고 적용해보는 것이 좋을 것 같다고 판단해 이것을 선택해보겠습니다.

sabgilhun commented on app/src/main/java/com/example/todo/airbnb/data/datasource/DateLocalDataSource.kt:
```diff
@@ -0,0 +1,61 @@
+import java.time.LocalDate
+import java.time.temporal.TemporalAdjusters
+
+typealias CalendarYear = List<CalendarMonth>
```
`typealias` 공식 문서를 보니 용도에 맞게 잘 사용된 것 같네요.
https://kotlinlang.org/docs/type-aliases.html

저도 이렇게는 사용 안해봤는데, 좋은 방식인 것 같습니다. 멋지네요. 배워갑니다 🥇

taewooyo replied:
감사합니다!!

sabgilhun commented on app/src/main/java/com/example/todo/airbnb/data/datasource/DateLocalDataSource.kt:
```diff
@@ -0,0 +1,61 @@
+@RequiresApi(Build.VERSION_CODES.O)
+class DatesLocalDataSource {
+
+    private val dayOfWeek = listOf(
```
네이밍 `monthOfYear` 이건 어떤가요?

그리고, Enum으로 정의해도 좋아 보입니다!

```kotlin
enum class MonthOfYear(val order: Int) {
    January(1)
    ...
}
```

taewooyo replied:
month는 정해져있는 (1월 ~12월) 달이기 때문에 enum으로 관리하는게 엄청 좋아보입니다. 적용하도록 하겠습니다!!

sabgilhun commented on app/src/main/java/com/example/todo/airbnb/data/datasource/DateLocalDataSource.kt:
```diff
@@ -0,0 +1,61 @@
+
+    private fun getDayList(): List<CalendarMonth> {
+        val list = mutableListOf<CalendarMonth>()
+        for (month in 1..12) {
```
```
dayOfWeek.forEachIndexed {

}

```kotlin
위와 같은 방식이나, enum을 적용 하셨다면 아래와 같이도 가능합니다.
```
private fun getDayList(): List<CalendarMonth> = MonthOfYear.values().map {
    val date = LocalDate.of(2021, it.order, 1)
    CalendarMonth(
        name = it.name,
        year = date.year.toString(),
        numDays = date.with(TemporalAdjusters.lastDayOfMonth()).dayOfMonth,
        monthNumber = it.order,
        startDayOfWeek = DayOfWeek.values(date.dayOfWeek.toString()) ?: DayOfWeek.Sunday
    )
}
```

sabgilhun commented on app/src/main/java/com/example/todo/airbnb/data/datasource/DateLocalDataSource.kt:
```diff
@@ -0,0 +1,61 @@
+        return list
+    }
+
+    @SuppressLint("VisibleForTests")
```
뭔가 실수로 `clearDates` 메소드에 `@VisibleForTesting` 어노테이션이 붙어서 워닝 지우려고 쓰신것 같네요.

`clearDates` 메소드위에 `@VisibleForTesting`를 지우면, @SuppressLint("VisibleForTests") 안 쓰셔도 됩니다~


참고로 `@VisibleForTesting`은 특정 클래스의 private 메서드 or private 필드를 테스트를 위해 접근하고 싶을때 사용합니다~

taewooyo replied:
수정했습니다!!

sabgilhun commented on app/src/main/java/com/example/todo/airbnb/presentation/search/date/components/CalendarMonth.kt:
```diff
@@ -0,0 +1,64 @@
+package com.example.todo.airbnb.presentation.search.date.components
+
+data class CalendarMonth(
+    val name: String,
+    val year: String,
+    val numDays: Int,
+    val monthNumber: Int,
+    val startDayOfWeek: DayOfWeek,
```
model 구조화가 잘 되어 있는 것 같아요. 👍 

사소하지만, 저는 components, 패키지가 ui component를 모아 두는 것이라 생각했거든요.

지금 `CalendarMonth`, `CalendarDay`은 model class라 이 class는 `date.model` 이런식의 패키지를 만들어서 따로 관리하는 게 맞지 않을까? 라는 생각이 드는데, 어떤가요?

taewooyo replied:
달력에 구현하는데 너무 신경쓴 나머지 component에서 개발하다가 옮겨야겠다라는 생각을 가지고 있었는데 놓친 것 같습니다. date layer의 model.date패키지 만들어서 사용하면 가독성도 좋을 것 같다고 판단됩니다.

sabgilhun commented on app/src/main/java/com/example/todo/airbnb/presentation/search/date/components/DatesSelectedState.kt:
```diff
@@ -0,0 +1,97 @@
+import androidx.compose.runtime.setValue
+import com.example.todo.airbnb.data.datasource.CalendarYear
+
+class DatesSelectedState(private val year: CalendarYear) {
```
UI state에 대한 정의를 한 model로 보이는데요.

제가 이해하고 있는 UI state 객체는, 말 그대로 UI 상태만 불변 객체로 갖고 있어야 한다고 생각해요.
https://developer.android.com/jetpack/guide/ui-layer?hl=ko#define-ui-state

`DatesSelectedState` 는 UI 상태 정보만 갖는 data class가 되어야 할것 같고, `daySelected` 함수 들이 ViewModel에 있어야 할 로직처럼 보이고요. ViewModel에 `State<DatesSelectedState>`를 갖는 구조가 되어야 하지 않을까 싶어요!

taewooyo replied:
UI State 별로 관리할 수 있도록 수정해보도록 하겠습니다.

sabgilhun reviewed (commented):
안녕하세요. 스티치, 정박 PR 잘 봤습니다.

Compose로 UI 구현하는 것이 샘플도 많지 않아서 힘들텐데, 완벽하게 구현하신 것 같아요. 최고입니다.
그래프쪽도 컴포즈만으로 멋지게 구현하신것 같아요. 👍 

------------------------
## 질문 답변
### 그래프 거꾸로 되는 문제
아마 0,0 좌표가 왼쪽 제일 위라서... 저게 맞는것 같아요. 어떻게 잘 뒤집는 방법을 찾으셨네요. ㅎㅎ
Compose로 UI 구성하실때 보면 따로 지정안하면 왼쪽 위에 UI 생기잖아요? 같은 이유라고 생각합니다.

taewooyo committed:
feat: @RequiresApi 해결

@RequiresApi 해결
- 디슈가링으로 해결완료

taewooyo committed:
feat: 1~12월 관리를 위한 Enum class 설정

기존의 1~12월 관리하는 list 수정
- 1~12을 관리하는 Enum class로 개선
- 사용자에게 보기 편하게 수정하기 위해 글자 포맷팅

taewooyo committed:
feat: 파일 관리 및 UI State 관리

객체 분리
- 객체들의 의존성 분리 및 패키지 별 관리

UI State 관리
- UI State 관리르 위해 ViewModel에 변경 관리로 개선

taewooyo committed:
Merge pull request #35 from bn-tw2020/feature/feedback-33

[feat] 2주차 수요일 피드백(33)

taewooyo commented:
꼼꼼하게 피드백 해주셔서 정말 감사합니다. @sabgilhun 이 말씀해주신, 권유해주신 방법을 참고하여 피드백을 반영했습니다.
현재 Pull Request에 포함했습니다. 확인해주시면 감사하겠습니다.

사소한 부분까지 체크해주시면 리뷰를 해주시니깐, 저의 부족함도 알게되니 좋은 것 같습니다!!!
